src/box/memtx_engine.cc

```diff
@@ -282,7 +293,7 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	 * Snapshot entries are ordered by the space id, it means that if there
 	 * are no spaces, then all system spaces are definitely missing.
 	 */
-	if (is_space_system < 0)
+	if (!any_system_space)
```

Comment:
I don't understand: does this mean that if we recovered at least one system space, we assume everything's fine? I thought we wanted to ensure that we recovered at least one record unrelated to system spaces.

Anyway, it'd probably be better to move the flag update completely to `memtx_engine_recovery_snapshot_row`. I mean something like this:

<details><summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 8fdcb1b78ea6..001a5467687d 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -218,7 +218,8 @@ memtx_engine_shutdown(struct engine *engine)
 
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
-				  struct xrow_header *row, int *is_space_system);
+				  struct xrow_header *row,
+				  bool *recovered_system_spaces);
 
 int
 memtx_engine_recover_snapshot(struct memtx_engine *memtx,
@@ -238,16 +239,16 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	int rc;
 	struct xrow_header row;
 	uint64_t row_count = 0;
-	int is_space_system = -1;
 	bool force_recovery = false;
+	bool recovered_system_spaces = false;
 	/*
 	 * In case when we read system space, we can't ignore errors.
 	 */
 	while ((rc = xlog_cursor_next(&cursor, &row, force_recovery)) == 0) {
 		row.lsn = signature;
-		rc = memtx_engine_recover_snapshot_row(memtx, &row,
-						       &is_space_system);
-		force_recovery = is_space_system == 0 ?
+		rc = memtx_engine_recover_snapshot_row(
+			memtx, &row, &recovered_system_spaces);
+		force_recovery = recovered_system_spaces ?
 				 memtx->force_recovery : false;
 		if (rc < 0) {
 			if (!force_recovery)
@@ -282,7 +283,7 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	 * Snapshot entries are ordered by the space id, it means that if there
 	 * are no spaces, then all system spaces are definitely missing.
 	 */
-	if (is_space_system < 0)
+	if (!recovered_system_spaces)
 		panic("snapshot `%s' has no system spaces", cursor.name);
 
 	return 0;
@@ -317,10 +318,12 @@ memtx_engine_recover_synchro(const struct xrow_header *row)
 
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
-				  struct xrow_header *row, int *is_space_system)
+				  struct xrow_header *row,
+				  bool *recovered_system_spaces)
 {
 	assert(row->bodycnt == 1); /* always 1 for read */
 	if (row->type != IPROTO_INSERT) {
+		*recovered_system_spaces = true;
 		if (row->type == IPROTO_RAFT)
 			return memtx_engine_recover_raft(row);
 		if (row->type == IPROTO_RAFT_PROMOTE)
@@ -333,7 +336,9 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	RegionGuard region_guard(&fiber()->gc);
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
-	*is_space_system = (request.space_id < BOX_SYSTEM_ID_MAX);
+	if (reqeust.space_id < BOX_SYSTEM_ID_MIN ||
+	    request.space_id > BOX_SYSTEM_ID_MAX)
+		*recovered_system_spaces = true;
 	struct space *space = space_cache_find(request.space_id);
 	if (space == NULL)
 		goto log_request;
```

</details>

Reply:
> does this mean that if we recovered at least one system space, we assume everything's fine

No, this flag is needed for the cases we don't find any system spaces in the snapshot, please see @Gumix's patch e0a9aed4.

Reply:
Thanks for the idea, the third flag is indeed redundant, applied your suggestion!

Reply:
Again, your patch doesn't work — there can be a snapshot that only has user spaces, and storage initialization will not fail. We need a separate flag to track if we recovered a system space.

Reply:
A enum and a flag still look ugly. I think it's possible to use only enum. What about something like this?

<details><summary>diff</summary>

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 8fdcb1b78ea6..43bc4234c278 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -216,9 +216,16 @@ memtx_engine_shutdown(struct engine *engine)
 	free(memtx);
 }
 
+enum snapshot_recovery_phase {
+	SNAPSHOT_RECOVERY_NOT_STARTED,
+	RECOVERING_SYSTEM_SPACES,
+	DONE_RECOVERING_SYSTEM_SPACES,
+};
+
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
-				  struct xrow_header *row, int *is_space_system);
+				  struct xrow_header *row,
+				  enum snapshot_recovery_phase *phase);
 
 int
 memtx_engine_recover_snapshot(struct memtx_engine *memtx,
@@ -238,17 +245,16 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	int rc;
 	struct xrow_header row;
 	uint64_t row_count = 0;
-	int is_space_system = -1;
 	bool force_recovery = false;
+	bool snapshot_recovery_phase phase = SNAPSHOT_RECOVERY_NOT_STARTED;
 	/*
 	 * In case when we read system space, we can't ignore errors.
 	 */
 	while ((rc = xlog_cursor_next(&cursor, &row, force_recovery)) == 0) {
 		row.lsn = signature;
-		rc = memtx_engine_recover_snapshot_row(memtx, &row,
-						       &is_space_system);
-		force_recovery = is_space_system == 0 ?
-				 memtx->force_recovery : false;
+		rc = memtx_engine_recover_snapshot_row(memtx, &row, &phase);
+		if (phase >= DONE_RECOVERING_SYSTEM_SPACES)
+			force_recovery = true;
 		if (rc < 0) {
 			if (!force_recovery)
 				break;
@@ -277,14 +283,6 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 		else
 			say_error("snapshot `%s' has no EOF marker", cursor.name);
 	}
-
-	/*
-	 * Snapshot entries are ordered by the space id, it means that if there
-	 * are no spaces, then all system spaces are definitely missing.
-	 */
-	if (is_space_system < 0)
-		panic("snapshot `%s' has no system spaces", cursor.name);
-
 	return 0;
 }
 
@@ -315,12 +313,26 @@ memtx_engine_recover_synchro(const struct xrow_header *row)
 	return txn_limbo_process(&txn_limbo, &req);
 }
 
+static inline int
+complete_system_space_recovery(enum snapshot_recovery_phase *phase)
+{
+	if (*phase < RECOVERING_SYSTEM_SPACES) {
+		diag_set(ClientError, ER_MISSING_SYSTEM_SPACES);
+		return -1;
+	}
+	*phase = DONE_RECOVERING_SYSTEM_SPACES;
+	return 0;
+}
+
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
-				  struct xrow_header *row, int *is_space_system)
+				  struct xrow_header *row,
+				  enum snapshot_recovery_phase *phase)
 {
 	assert(row->bodycnt == 1); /* always 1 for read */
 	if (row->type != IPROTO_INSERT) {
+		if (complete_system_space_recovery(phase) != 0)
+			return -1;
 		if (row->type == IPROTO_RAFT)
 			return memtx_engine_recover_raft(row);
 		if (row->type == IPROTO_RAFT_PROMOTE)
@@ -333,7 +345,9 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	RegionGuard region_guard(&fiber()->gc);
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
-	*is_space_system = (request.space_id < BOX_SYSTEM_ID_MAX);
+	if (!space_id_is_system(request.space_id) &&
+	    complete_system_space_recovery(phase) != 0)
+		return -1;
 	struct space *space = space_cache_find(request.space_id);
 	if (space == NULL)
 		goto log_request;
```

</details>